# Лабораторные работы по дисциплине "Методы, средства и технологии мультимедиа"

**Выполнила студентка гр. М8О-406Б-21 Волошинская Евгения Владимировна**

В качестве данных я выбрала датасет Semantic Segmentation Drone Dataset (https://www.kaggle.com/datasets/santurini/semantic-segmentation-drone-dataset).
Эти данные можно использовать для обучения моделей компьютерного зрения, которые автоматически анализируют аэрофотосъёмку с дронов: прокладывают безопасный маршрут, обновляют карты дорог и зданий, оценивают состояние инфраструктуры и выделяют зелёные зоны. Это повышает точность навигации, снижает ручную работу операторов и ускоряет принятие решений в городских службах.

**Датасет поддерживает две постановки задачи семантической сегментации:**

☑ Бинарная – отделить объекты от фона (быстрый базовый вариант).

☐ 5-классная – раскрасить каждый пиксель по макрогруппам: небо, здания/препятствия, дорога/земля, растительность, транспорт/люди.

Была выбрана задача бинарной сегментации, т.к. в реализации проекта используются ограниченные возможности бесплатной версии Google Colab.

**Ключевые метрики качества:**

• mIoU (mean Intersection over Union)– основной индикатор перекрытия предсказанных и истинных масок,
где IoU - отношение пересечения предсказанной и истинной масок к их объединению, mIoU - среднее значение IoU по всем классамю

• Dice/F1-score – гармоника точности и полноты по пикселям, особенно важна при бинарной задаче.

• Pixel Accuracy – доля правильно размеченных пикселей (вспомогательная).

IoU и Dice по каждому классу – показывают, не «теряются» ли редкие объекты.

####Структура датасета

В основной папке датасета находится папка binary_dataset/binary_dataset, которая содержит две подпапки: с оригинальными изображениями (original_images) и с размеченными данными (images_semantic)
```
binary_dataset/
├── original_images/
│   ├── 000.png
│   └── 001.png
└── images_semantic/
    ├── 000.png
    └── 001.png
```

## Лабораторная работа №7: Проведение исследований моделями семантической сегментации

In [3]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("santurini/semantic-segmentation-drone-dataset")

print("Path to dataset files:", path)

Path to dataset files: /kaggle/input/semantic-segmentation-drone-dataset


In [2]:
!pip install -q --upgrade pip
!pip install -q pytorch-lightning segmentation_models_pytorch albumentations timm tqdm

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 28.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.1/823.1 kB 26.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 48.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 187.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 187.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 39.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 22.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 71.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 80.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 68.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 52.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 206.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

####Бейзлайн

В качестве бейзлайна были обучены две модели:

1. Сверточная модель U-Net с энкодером ResNet34;

2. DeepLabV3+ с трансформерным энкодером MIT-B0.

Оценка производилась по метрикам IoU (Intersection over Union), Accuracy и Dice Loss на тестовой выборке.

In [ ]:
# ================================================================
# 1. Подключение библиотек
# ================================================================
import os, random, cv2, numpy as np, torch, pytorch_lightning as pl
from pathlib import Path
import albumentations as A
from albumentations.pytorch import ToTensorV2
import segmentation_models_pytorch as smp
import inspect

USE_WANDB = False
SEED = 42
random.seed(SEED); np.random.seed(SEED); torch.manual_seed(SEED)

# ================================================================
# 2. Пути к папкам с файлами
# ================================================================
ROOT = Path("/kaggle/input/semantic-segmentation-drone-dataset/binary_dataset/binary_dataset")
IMG_DIR  = ROOT / "original_images"
MASK_DIR = ROOT / "images_semantic"
assert IMG_DIR.exists() and MASK_DIR.exists()

# ================================================================
# 3. FILES + SPLIT
# ================================================================
imgs  = sorted(IMG_DIR.glob("*.png"))
if not imgs:
    raise RuntimeError("В original_images нет *.png файлов")

masks = [MASK_DIR / f"{p.stem}.png" for p in imgs]
for m in masks:
    if not m.exists():
        raise RuntimeError(f"Нет маски для {m.stem}.png")

perm  = np.random.permutation(len(imgs))
tr, vl = int(.8*len(perm)), int(.9*len(perm))
split_files = dict(
    train=(np.array(imgs)[perm[:tr]],   np.array(masks)[perm[:tr]]),
    val  =(np.array(imgs)[perm[tr:vl]], np.array(masks)[perm[tr:vl]]),
    test =(np.array(imgs)[perm[vl:]],   np.array(masks)[perm[vl:]]),
)

# ================================================================
# 4. DATASET
# ================================================================
class DroneBinDS(torch.utils.data.Dataset):
    def __init__(self, imgs, masks, tfm):
        self.imgs, self.masks, self.t = list(imgs), list(masks), tfm
    def __len__(self):  return len(self.imgs)
    def __getitem__(self, i):
        img  = cv2.cvtColor(cv2.imread(str(self.imgs[i])), cv2.COLOR_BGR2RGB)
        mask = cv2.imread(str(self.masks[i]), 0)
        mask = (mask > 127).astype("float32")
        batch = self.t(image=img, mask=mask)
        return batch["image"], batch["mask"].unsqueeze(0)

# ================================================================
# 5. AUGMENTATIONS
# ================================================================
train_tf = A.Compose([
    A.Resize(736, 736),
    A.Normalize(), ToTensorV2()
])
test_tf  = train_tf

def run_baseline():
    for name, lr in [("Unet",3e-4), ("DeepLabV3Plus",5e-4)]:
        trainer = pl.Trainer(max_epochs=15,
                             accelerator="gpu" if torch.cuda.is_available() else "cpu",
                             devices=1, log_every_n_steps=10,
                             enable_checkpointing=False)
        trainer.fit(LitSeg(name, lr)); trainer.test(LitSeg(name, lr))
# ================================================================
# 6. LIGHTNING:
# PyTorch Lightning — высокоуровневая обёртка над PyTorch, которая позволяет писать более чистый, модульный и удобный для экспериментов код.
# ================================================================
MODELS = {
    "Unet": lambda: smp.Unet("resnet34", encoder_weights="imagenet",
                             in_channels=3, classes=1, activation=None),
    "DeepLabV3Plus": lambda: smp.DeepLabV3Plus(
        encoder_name="mit_b0", encoder_weights="imagenet",
        in_channels=3, classes=1, activation=None),
}
LOSS = smp.losses.DiceLoss("binary", from_logits=True)

class LitSeg(pl.LightningModule):
    def __init__(self, name, lr=3e-4, bs=8):
        super().__init__()
        self.model, self.lr, self.bs = MODELS[name](), lr, bs
        self.save_hyperparameters()
    # data
    def setup(self, stage=None):
        self.ds_train = DroneBinDS(*split_files["train"], train_tf)
        self.ds_val   = DroneBinDS(*split_files["val"],   test_tf)
        self.ds_test  = DroneBinDS(*split_files["test"],  test_tf)
    def train_dataloader(self):
        return torch.utils.data.DataLoader(self.ds_train,
                                           batch_size=self.bs,
                                           shuffle=True,
                                           num_workers=2)
    def val_dataloader(self):
        return torch.utils.data.DataLoader(self.ds_val,
                                           batch_size=self.bs,
                                           shuffle=False,
                                           num_workers=2)
    def test_dataloader(self):
        return torch.utils.data.DataLoader(self.ds_test,
                                           batch_size=self.bs,
                                           shuffle=False,
                                           num_workers=2)
    # loop
    def shared_step(self,b,tag):
        x,y=b
        logits=self.model(x); loss=LOSS(logits,y)
        pred=(torch.sigmoid(logits)>=.5).long()
        tp,fp,fn,tn=smp.metrics.get_stats(pred,y.long(),mode="binary")
        iou=smp.metrics.iou_score(tp,fp,fn,tn,reduction="micro-imagewise")
        acc=smp.metrics.accuracy(tp,fp,fn,tn,reduction="micro")
        self.log_dict({f"{tag}_loss":loss,f"{tag}_iou":iou,f"{tag}_acc":acc},prog_bar=True)
        return loss
    def training_step(self,b,_):   return self.shared_step(b,"train")
    def validation_step(self,b,_): return self.shared_step(b,"val")
    def test_step(self,b,_):       return self.shared_step(b,"test")
    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.lr)

# ================================================================
# 7. RUN
# ================================================================
def run(name, lr):
    logger = (pl.loggers.WandbLogger(project="seg_drone", name=name)
              if USE_WANDB else None)
    model  = LitSeg(name, lr)
    trainer = pl.Trainer(accelerator="gpu" if torch.cuda.is_available() else "cpu",
                         devices=1, max_epochs=15, logger=logger,
                         log_every_n_steps=10, enable_checkpointing=False)
    trainer.fit(model); trainer.test(model)

run("Unet",          lr=3e-4)   # CNN
run("DeepLabV3Plus", lr=5e-4)   # Transformer (mit_b0)

INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name  | Type | Params | Mode 
---------------------------------------
0 | model | Unet | 24.4 M | train
---------------------------------------
24.4 M    Trainable params
0         Non-trainable params
24.4 M    Total params
97.745    Total estimated model params size (MB)
188       Modules in train mode
0         Modules in eval mode


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=15` reached.
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_acc          │    0.9330824613571167     │
│         test_iou          │    0.7561880946159363     │
│         test_loss         │    0.13053195178508759    │
└───────────────────────────┴───────────────────────────┘

config.json:   0%|          | 0.00/135 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/14.3M [00:00<?, ?B/s]

INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name  | Type          | Params | Mode 
------------------------------------------------
0 | model | DeepLabV3Plus | 4.1 M  | train
------------------------------------------------
4.1 M     Trainable params
0         Non-trainable params
4.1 M     Total params
16.544    Total estimated model params size (MB)
231       Modules in train mode
0         Modules in eval mode


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=15` reached.
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_acc          │    0.9262682795524597     │
│         test_iou          │    0.7268131971359253     │
│         test_loss         │    0.14089824259281158    │
└───────────────────────────┴───────────────────────────┘

**Интерпретация результатов Бейзлайна:**

**Модель 1: U-Net (с энкодером ResNet34)**

*   **Параметры модели:**
    *   Общее количество параметров: 24.4 миллиона.
    *   Размер модели: ~97.7 MB.
    *   Это относительно крупная модель для U-Net с таким энкодером, что говорит о ее потенциальной способности к хорошему обучению на сложных данных.

**Модель 2: DeepLabV3+ (с энкодером MIT-B0 - трансформерным)**

*   **Параметры модели:**
    *   Общее количество параметров: 4.1 миллиона.
    *   Размер модели: ~16.5 MB.
    *   Эта модель значительно легче (меньше параметров), чем U-Net с ResNet34. Это важно для развертывания на устройствах с ограниченными ресурсами.

---

**Вывод:**
1.  **U-Net с энкодером ResNet34 (сверточная архитектура):**
    Модель имеет 24.4 миллиона параметров. Обучение проводилось в течение 15 эпох с использованием функции потерь DiceLoss и оптимизатора Adam. На последней эпохе обучения были достигнуты следующие показатели на валидационной выборке: `val_loss` = 0.125, `val_iou` = 0.772, `val_acc` = 0.931.
    При оценке на тестовой выборке модель U-Net показала следующие результаты:
    *   **Test IoU (mIoU): 0.756**
    *   Test Pixel Accuracy: 0.933
    *   Test Loss: 0.131

2.  **DeepLabV3+ с энкодером MIT-B0 (архитектура с трансформерным энкодером):**
    Данная модель является более легковесной, имея 4.1 миллиона параметров. Условия обучения были аналогичны модели U-Net (15 эпох, DiceLoss, Adam). На последней эпохе обучения показатели на валидационной выборке составили: `val_loss` = 0.138, `val_iou` = 0.741, `val_acc` = 0.922.
    Результаты на тестовой выборке для модели DeepLabV3+ с энкодером MIT-B0:
    *   **Test IoU (mIoU): 0.727**
    *   Test Pixel Accuracy: 0.926
    *   Test Loss: 0.141

**Таким образом,**
 обе модели продемонстрировали способность к решению задачи бинарной семантической сегментации на выбранном датасете. Сверточная модель U-Net с энкодером ResNet34 показала несколько лучшие результаты по основной метрике IoU (0.756 против 0.727) на тестовой выборке по сравнению с моделью DeepLabV3+ с трансформерным энкодером MIT-B0. Однако, U-Net является значительно более тяжелой моделью по количеству параметров. Скорость обучения U-Net также была выше (около 41 секунды на эпоху против 55 секунд для DeepLabV3+). ба подхода продемонстрировали высокое качество сегментации, однако классическая сверточная архитектура оказалась более эффективной на данном датасете. Полученные значения IoU (72-75%) являются хорошей отправной точкой для дальнейших улучшений в рамках пункта 3 лабораторной работы.

####Улучшенный бейзлайн

ГИПОТЕЗЫ

**H1: Более сильные аугментации (пространственные и цветовые)**

Для повышения обобщающей способности моделей были применены расширенные аугментации изображений, включающие как пространственные (например, случайные обрезки, повороты, масштабирование), так и цветовые преобразования (изменение яркости, контраста, добавление шума и размытия). Это позволяет модели лучше справляться с разнообразием входных данных и уменьшает переобучение.


**H2: Более мощные энкодеры (ResNet50 для U-Net, mit_b2 для DeepLabV3+)**

Были использованы более глубокие и производительные энкодеры: ResNet50 для U-Net и mit_b2 (Vision Transformer из библиотеки timm) для DeepLabV3+. Это увеличивает способность модели извлекать сложные признаки из изображений, что особенно важно для задач сегментации с высоким разрешением и сложной структурой объектов.

**H3: Комбинированная функция потерь (Dice + BCE)**

В качестве функции потерь использовалась комбинация Dice Loss и Binary Cross-Entropy (BCE) с весами 0.7 и 0.3 соответственно. Dice Loss хорошо подходит для задач сегментации, так как напрямую оптимизирует перекрытие масок, а BCE способствует более стабильному обучению и учитывает пиксельную точность. Совместное использование этих функций позволяет достичь лучшего баланса между качеством сегментации и стабильностью обучения.

**H4: Продвинутый планировщик скорости обучения (PolyLR) + AdamW**

Для оптимизации обучения был выбран оптимизатор AdamW с weight decay, который помогает бороться с переобучением. В качестве планировщика скорости обучения использовался PolyLR (polynomial learning rate decay), который плавно уменьшает learning rate по мере обучения, что способствует более эффективной и стабильной сходимости модели.

In [ ]:
print(f"\n{'='*15} ЗАПУСК УЛУЧШЕННОГО БЕЙЗЛАЙНА {'='*15}")
# ================================================================
# УЛУЧШЕННЫЕ АУГМЕНТАЦИИ, H1
# ================================================================
H, W = 736, 736

train_tf_improved = A.Compose([
    make_rrc(H, W, scale=(0.5, 1.0), p=1.0), # ИСПРАВЛЕНО: scale до 1.0
    A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.1, rotate_limit=15, p=0.7),
    A.HorizontalFlip(p=0.5),
    A.VerticalFlip(p=0.5),
    A.RandomBrightnessContrast(brightness_limit=0.3, contrast_limit=0.3, p=0.5),
    A.GaussianBlur(blur_limit=(3, 7), p=0.3),
    A.GaussNoise(var_limit=(10.0, 50.0), p=0.3),
    A.Normalize(),
    ToTensorV2(),
])

test_tf_improved = A.Compose([
    make_resize(H, W),
    A.Normalize(),
    ToTensorV2()
])

# ================================================================
# LIGHTNING MODULE С УЛУЧШЕНИЯМИ
# ================================================================
MODELS_IMP = {
    "Unet_resnet50": lambda: smp.Unet("resnet50", encoder_weights="imagenet", # H2
                                      in_channels=3, classes=1, activation=None),
    "DeepLabV3Plus_mit_b2": lambda: smp.DeepLabV3Plus( # H2
        encoder_name="mit_b2", encoder_weights="imagenet",
        in_channels=3, classes=1, activation=None),
}

# H3: Комбинированная функция потерь
dice_loss_imp = smp.losses.DiceLoss("binary", from_logits=True)
bce_loss_imp = torch.nn.BCEWithLogitsLoss()
def combo_loss_fn(logits, target):
    return 0.7 * dice_loss_imp(logits, target) + 0.3 * bce_loss_imp(logits, target)

# H4: PolyLR планировщик
def poly_lr_scheduler_fn(step, max_steps, initial_lr_factor, power=0.9):
    if step >= max_steps: # Prevent going to 0 or negative if max_steps is approximate
        return 1e-6 / initial_lr_factor # Return a very small factor instead of 0
    return initial_lr_factor * (1 - step / max_steps) ** power


class LitSegImproved(pl.LightningModule):
    def __init__(self, name, lr=3e-4, bs=8, total_epochs=15):
        super().__init__()
        self.model = MODELS_IMP[name]()
        self.initial_lr = lr
        self.bs = bs
        self.total_epochs = total_epochs
        self.save_hyperparameters()

    # Data
    def setup(self, stage=None):
        # Используем DroneBinDS из бейзлайна
        self.ds_train_imp = DroneBinDS(*split_files["train"], train_tf_improved)
        self.ds_val_imp   = DroneBinDS(*split_files["val"],   test_tf_improved)
        self.ds_test_imp  = DroneBinDS(*split_files["test"],  test_tf_improved)

    def train_dataloader(self):
        return torch.utils.data.DataLoader(self.ds_train_imp, batch_size=self.bs, shuffle=True, num_workers=2, pin_memory=True)
    def val_dataloader(self):
        return torch.utils.data.DataLoader(self.ds_val_imp, batch_size=self.bs, shuffle=False, num_workers=2, pin_memory=True)
    def test_dataloader(self):
        return torch.utils.data.DataLoader(self.ds_test_imp, batch_size=self.bs, shuffle=False, num_workers=2, pin_memory=True)

    # Loop
    def shared_step(self, batch, tag):
        x,y = batch
        logits = self.model(x)
        loss   = combo_loss_fn(logits, y) # H3
        pred   = (torch.sigmoid(logits) >= .5).long()
        tp, fp, fn, tn = smp.metrics.get_stats(pred, y.long(), mode="binary")
        iou = smp.metrics.iou_score(tp, fp, fn, tn, reduction="micro-imagewise")
        acc = smp.metrics.accuracy(tp, fp, fn, tn, reduction="micro")
        self.log_dict({f"{tag}_loss":loss, f"{tag}_iou":iou, f"{tag}_acc":acc}, prog_bar=True, on_step=(tag=="train"), on_epoch=True)
        if tag == "train":
            current_lr = self.lr_schedulers().get_last_lr()[0]
            self.log("learning_rate", current_lr, on_step=True, on_epoch=False, prog_bar=False)
        return loss

    def training_step(self,b,_):   return self.shared_step(b,"train")
    def validation_step(self,b,_): return self.shared_step(b,"val")
    def test_step(self,b,_):       return self.shared_step(b,"test")

    def configure_optimizers(self):
        optimizer = torch.optim.AdamW(self.parameters(), lr=self.initial_lr, weight_decay=1e-5) # H4: AdamW

        # Расчет общего количества шагов для PolyLR
        # Оцениваем по длине датасета, так как trainer.num_training_batches еще не доступен здесь
        num_train_samples = len(split_files["train"][0])
        num_batches_per_epoch = (num_train_samples // self.bs) + (1 if num_train_samples % self.bs != 0 else 0)

        total_training_steps = num_batches_per_epoch * self.total_epochs

        scheduler = torch.optim.lr_scheduler.LambdaLR(
            optimizer,
            lr_lambda=lambda step: poly_lr_scheduler_fn(step, total_training_steps, 1.0, power=0.9)
        )
        return {"optimizer": optimizer, "lr_scheduler": {"scheduler": scheduler, "interval": "step"}}


# ================================================================
# TRAIN & TEST УЛУЧШЕННОГО БЕЙЗЛАЙНА
# ================================================================
def run_improved(name, lr, total_epochs=15, batch_size=8):
    logger = (pl.loggers.WandbLogger(project="seg_drone_improved", name=name)
              if USE_WANDB else True)

    m = LitSegImproved(name, lr=lr, total_epochs=total_epochs, bs=batch_size)

    trainer = pl.Trainer(
        accelerator="gpu" if torch.cuda.is_available() else "cpu",
        devices=1,
        max_epochs=total_epochs,
        logger=logger,
        log_every_n_steps=10,
        enable_checkpointing=False
    )
    trainer.fit(m)
    trainer.test(m)

# Запускаем эксперименты для улучшенного бейзлайна
EPOCHS_IMPROVED = 15
BATCH_SIZE_IMPROVED = 8

print(f"\nЗапуск обучения улучшенных моделей на {EPOCHS_IMPROVED} эпох, batch_size={BATCH_SIZE_IMPROVED}...")
run_improved("Unet_resnet50", lr=1e-4, total_epochs=EPOCHS_IMPROVED, batch_size=BATCH_SIZE_IMPROVED)
run_improved("DeepLabV3Plus_mit_b2", lr=2e-4, total_epochs=EPOCHS_IMPROVED, batch_size=BATCH_SIZE_IMPROVED)

print(f"\n{'='*15} ОБУЧЕНИЕ УЛУЧШЕННОГО БЕЙЗЛАЙНА ЗАВЕРШЕНО {'='*15}")


=============== ЗАПУСК УЛУЧШЕННОГО БЕЙЗЛАЙНА ===============

Запуск обучения улучшенных моделей на 15 эпох, batch_size=8...


<ipython-input-18-6552a8870382>:28: UserWarning: Argument(s) 'var_limit' are not valid for transform GaussNoise
  A.GaussNoise(var_limit=(10.0, 50.0), p=0.3),
INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name  | Type | Params | Mode 
---------------------------------------
0 | model | Unet | 32.5 M | train
---------------------------------------
32.5 M    Trainable params
0         Non-trainable params
32.5 M    Total params
130.084   Total estimated model params size (MB)
223       Modules in train mode
0         Modules in eval mode


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=15` reached.
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]


Testing: |          | 0/? [00:00<?, ?it/s]

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_acc          │    0.9254651069641113     │
│         test_iou          │    0.7329956293106079     │
│         test_loss         │    0.2021360844373703     │
└───────────────────────────┴───────────────────────────┘

config.json:   0%|          | 0.00/135 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/98.9M [00:00<?, ?B/s]

INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name  | Type          | Params | Mode 
------------------------------------------------
0 | model | DeepLabV3Plus | 25.3 M | train
------------------------------------------------
25.3 M    Trainable params
0         Non-trainable params
25.3 M    Total params
101.396   Total estimated model params size (MB)
383       Modules in train mode
0         Modules in eval mode


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

OutOfMemoryError: CUDA out of memory. Tried to allocate 1.07 GiB. GPU 0 has a total capacity of 14.74 GiB of which 698.12 MiB is free. Process 2724 has 14.06 GiB memory in use. Of the allocated memory 13.79 GiB is allocated by PyTorch, and 135.54 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

Ошибка CUDA out of memory для второй модели (DeepLabV3Plus_mit_b2) на бесплатном GPU Google Colab Tesla T4 с ~15 ГБ VRAM довольно ожидаема, так как полученные модели тяжеловесны, а трансформерные энкодеры, даже "легкие" как mit_b2, могут быть более требовательны к памяти, чем сверточные (ResNet50). Попробуем решить данную проблему. Перезапустим среду выполнения.

In [ ]:
# ================================================================
# 0. INSTALL DEPENDENCIES
# ================================================================
!pip install -q --upgrade pip
!pip install -q pytorch-lightning segmentation_models_pytorch albumentations timm tqdm

# ================================================================
# 1. IMPORTS & ENVIRONMENT SETUP
# ================================================================
import os
os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True' # Попытка уменьшить фрагментацию

import random
import cv2
import numpy as np
import torch
import pytorch_lightning as pl
from pathlib import Path
import albumentations as A
from albumentations.pytorch import ToTensorV2
import segmentation_models_pytorch as smp
import inspect # Для универсальных аугментаций
import gc      # Для сборщика мусора
USE_WANDB = False
SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed_all(SEED)

In [ ]:
# ================================================================
# 2. PATHS TO DATASET (на Google Drive)
# ================================================================
DRIVE_ROOT = Path("/content/drive/MyDrive/Colab Notebooks/2/binary_dataset/binary_dataset")
# Локальная папка на диске Colab для копирования данных
LOCAL_ROOT = Path("/content/dataset_local_binary_seg")

IMG_DIR_DRIVE  = DRIVE_ROOT / "original_images"
MASK_DIR_DRIVE = DRIVE_ROOT / "images_semantic"

IMG_DIR_LOCAL  = LOCAL_ROOT / "original_images"
MASK_DIR_LOCAL = LOCAL_ROOT / "images_semantic"

# ================================================================
# 2.1 COPY DATASET TO LOCAL COLAB DISK (ВАЖНО ДЛЯ СКОРОСТИ И СТАБИЛЬНОСТИ)
# ================================================================
import shutil
if LOCAL_ROOT.exists():
    print(f"Локальная копия {LOCAL_ROOT} уже существует, используем ее.")
else:
    print(f"Копируем данные из {DRIVE_ROOT} в {LOCAL_ROOT}...")
    try:
        LOCAL_ROOT.mkdir(parents=True, exist_ok=True)
        items_to_copy = [item for item in DRIVE_ROOT.iterdir()]
        for item in items_to_copy:
            dest_path = LOCAL_ROOT / item.name
            if item.is_dir():
                shutil.copytree(item, dest_path)
            else:
                shutil.copy2(item, dest_path)

        print("Копирование датасета на локальный диск завершено.")
    except Exception as e:
        print(f"Ошибка при копировании датасета: {e}")
        print("Убедитесь, что Google Drive примонтирован и путь корректен.")
        print("Проверьте также права доступа к папке Drive.")
        IMG_DIR_LOCAL  = IMG_DIR_DRIVE
        MASK_DIR_LOCAL = MASK_DIR_DRIVE
        print("Продолжаем работу с данными напрямую с Google Drive (может быть очень медленно и нестабильно).")


assert IMG_DIR_LOCAL.exists() and MASK_DIR_LOCAL.exists(), \
    f"Проверьте путь к датасету: {IMG_DIR_LOCAL} или {MASK_DIR_LOCAL} не найдены после попытки копирования."

# ================================================================
# 3. COLLECT FILES + SPLIT (80/10/10)
# ================================================================
imgs  = sorted(list(IMG_DIR_LOCAL.glob("*.png")) +
               list(IMG_DIR_LOCAL.glob("*.jpg"))) # Включаем оба популярных формата

if not imgs:
    raise RuntimeError(f"В {IMG_DIR_LOCAL} не найдено изображений (*.png или *.jpg). Проверьте путь и содержимое.")

print(f"Найдено {len(imgs)} изображений.")

masks = [MASK_DIR_LOCAL / f"{p.stem}.png" for p in imgs]
existing_masks = [m_path for m_path in masks if m_path.exists()]
if len(existing_masks) != len(imgs):
     missing_masks = [m_path for m_path in masks if not m_path.exists()]
     print(f"Предупреждение: Найдено {len(imgs)} изображений, но только {len(existing_masks)} масок. Отсутствуют первые 5: {missing_masks[:5]}...")
     # Обновляем список изображений, чтобы соответствовать найденным маскам
     imgs = [IMG_DIR_LOCAL / f"{m.stem}{Path(imgs[0]).suffix}" for m in existing_masks]
masks = existing_masks

if len(imgs) < 3: # Минимум для сплита train/val/test
     raise RuntimeError(f"Недостаточно данных ({len(imgs)} пар изображение/маска) для разбиения на train/val/test.")

perm  = np.random.permutation(len(imgs))
tr_idx_end = int(.8*len(perm))
val_idx_end = int(.9*len(perm))

# Гарантия хотя бы одного элемента в каждом сплите (если данных мало)
tr_idx_end = max(1, tr_idx_end)
val_idx_end = max(tr_idx_end + 1, val_idx_end)
if val_idx_end >= len(imgs):
    val_idx_end = len(imgs) # Весь остаток идет в val/test

split_files = dict(
    train=(np.array(imgs)[perm[:tr_idx_end]],   np.array(masks)[perm[:tr_idx_end]]),
    val  =(np.array(imgs)[perm[tr_idx_end:val_idx_end]], np.array(masks)[perm[tr_idx_end:val_idx_end]]),
    test =(np.array(imgs)[perm[val_idx_end:]], np.array(masks)[perm[val_idx_end:]]),
)

if len(split_files["train"][0]) == 0: raise RuntimeError("Обучающая выборка пуста после разбиения.")
if len(split_files["val"][0]) == 0:   print("Предупреждение: Валидационная выборка пуста.")
if len(split_files["test"][0]) == 0:  print("Предупреждение: Тестовая выборка пуста.")

print(f"Разбиение: train={len(split_files['train'][0])}, val={len(split_files['val'][0])}, test={len(split_files['test'][0])} семплов.")


# ================================================================
# 4. DATASET CLASS (DroneBinDS) - Использует пути из split_files
# ================================================================
class DroneBinDS(torch.utils.data.Dataset):
    def __init__(self, imgs_paths, masks_paths, transform_fn):
        self.imgs_paths = list(imgs_paths)
        self.masks_paths = list(masks_paths)
        self.transform_fn = transform_fn

    def __len__(self):
        return len(self.imgs_paths)

    def __getitem__(self, i):
        img  = cv2.cvtColor(cv2.imread(str(self.imgs_paths[i])), cv2.COLOR_BGR2RGB)
        mask = cv2.imread(str(self.masks_paths[i]), 0) # Читаем как grayscale
        mask = (mask > 127).astype("float32") # Преобразуем в бинарную 0/1

        augmented = self.transform_fn(image=img, mask=mask)
        image_tensor = augmented["image"] # Ожидается tensor (C, H, W) от ToTensorV2
        mask_tensor  = augmented["mask"]  # Ожидается tensor (H, W) от ToTensorV2

        # Добавляем канал к маске (ожидается (1, H, W) для DiceLoss/BCEWithLogitsLoss)
        if mask_tensor.ndim == 2:
             mask_tensor = mask_tensor.unsqueeze(0)

        return image_tensor, mask_tensor.float() # Маска тоже float для лосса


# ================================================================
# 5. AUGMENTATIONS (Улучшенный вариант, совместимый с Albumentations)
# ================================================================
# Возвращаемся к исходному разрешению 736x736
H_IMG, W_IMG = 736, 736

# Функции make_resize_fn и make_rrc_fn
def make_resize_fn(h, w):
    sig_params = inspect.signature(A.Resize).parameters
    if "height" in sig_params and "width" in sig_params: return A.Resize(height=h, width=w)
    elif "size" in sig_params: return A.Resize(size=(h, w))
    else: raise TypeError("Не удалось определить API для A.Resize")

def make_rrc_fn(h, w, **kwargs):
    sig_params = inspect.signature(A.RandomResizedCrop).parameters
    if "height" in sig_params and "width" in sig_params: return A.RandomResizedCrop(height=h, width=w, **kwargs)
    elif "size" in sig_params: return A.RandomResizedCrop(size=(h, w), **kwargs)
    else: raise TypeError("Не удалось определить API для A.RandomResizedCrop")


train_transforms_improved = A.Compose([
    make_rrc_fn(H_IMG, W_IMG, scale=(0.5, 1.0), p=1.0),
    A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.1, rotate_limit=15, p=0.7),
    A.HorizontalFlip(p=0.5),
    A.VerticalFlip(p=0.5),
    A.RandomBrightnessContrast(brightness_limit=0.3, contrast_limit=0.3, p=0.5),
    A.GaussianBlur(blur_limit=(3, 7), p=0.3),
    A.GaussNoise(p=0.3),
    A.Normalize(), # Выполняет нормализацию
    ToTensorV2(), # Конвертирует в Tensor
])

test_transforms_improved = A.Compose([
    make_resize_fn(H_IMG, W_IMG),
    A.Normalize(), # Выполняет нормализацию
    ToTensorV2() # Конвертирует в Tensor
])


# ================================================================
# 6. LIGHTNING MODULE (LitSegImproved)
# ================================================================
# Используем только модель DeepLabV3Plus с mit_b2
MODELS_TO_TRAIN = {
    "DeepLabV3Plus_mit_b2": lambda: smp.DeepLabV3Plus(
        encoder_name="mit_b2", encoder_weights="imagenet",
        in_channels=3, classes=1, activation=None),
}

# Комбинированная функция потерь
dice_loss_fn = smp.losses.DiceLoss("binary", from_logits=True)
bce_loss_fn = torch.nn.BCEWithLogitsLoss()
def combined_loss(logits, target):
    return 0.7 * dice_loss_fn(logits, target) + 0.3 * bce_loss_fn(logits, target)

# PolyLR планировщик
def poly_lr_scheduler(step, max_steps, initial_lr_factor, power=0.9):
    if step >= max_steps:
        return 1e-7 # Очень маленькое значение вместо 0, чтобы не обнулять совсем
    return initial_lr_factor * (1 - step / max_steps) ** power

class LitSegImproved(pl.LightningModule):
    def __init__(self, model_key_name, initial_lr=3e-4, batch_size=8, num_total_epochs=15):
        super().__init__()
        self.model = MODELS_TO_TRAIN[model_key_name]()
        self.initial_lr = initial_lr
        self.batch_size = batch_size
        self.num_total_epochs = num_total_epochs
        self.save_hyperparameters()

    def setup(self, stage=None):
        self.train_dataset = DroneBinDS(*split_files["train"], train_transforms_improved)
        self.val_dataset   = DroneBinDS(*split_files["val"],   test_transforms_improved)
        self.test_dataset  = DroneBinDS(*split_files["test"],  test_transforms_improved)

    def train_dataloader(self):
        # Увеличен num_workers до 4 для потенциального ускорения загрузки с локального диска
        return torch.utils.data.DataLoader(self.train_dataset, batch_size=self.batch_size, shuffle=True, num_workers=4, pin_memory=True)
    def val_dataloader(self):
        return torch.utils.data.DataLoader(self.val_dataset, batch_size=self.batch_size, shuffle=False, num_workers=4, pin_memory=True)
    def test_dataloader(self):
        return torch.utils.data.DataLoader(self.test_dataset, batch_size=self.batch_size, shuffle=False, num_workers=4, pin_memory=True)

    def shared_step(self, batch, stage_name):
        images, true_masks = batch
        logits = self.model(images)
        loss   = combined_loss(logits, true_masks)

        pred_masks = (torch.sigmoid(logits) >= 0.5).long()
        true_masks_long = true_masks.long()

        tp, fp, fn, tn = smp.metrics.get_stats(pred_masks, true_masks_long, mode="binary")
        # reduction="micro" - усреднение по батчу (и по изображениям внутри батча)
        iou_score = smp.metrics.iou_score(tp, fp, fn, tn, reduction="micro")
        accuracy_score = smp.metrics.accuracy(tp, fp, fn, tn, reduction="micro")

        # log_dict автоматически добавляет префикс stage_name
        self.log_dict({f"{stage_name}_loss":loss, f"{stage_name}_iou":iou_score, f"{stage_name}_acc":accuracy_score},
                      on_step=(stage_name=="train"), on_epoch=True, prog_bar=True)

        if stage_name == "train":
            # Get LR from the first optimizer's first param group
            if self.trainer.optimizers and len(self.trainer.optimizers) > 0:
                current_lr = self.trainer.optimizers[0].param_groups[0]['lr']
                self.log("learning_rate", current_lr, on_step=True, on_epoch=False, prog_bar=False)
            # else: LR scheduler might not be set up yet for the very first step

        return loss

    def training_step(self, batch_data, batch_idx):   return self.shared_step(batch_data,"train")
    def validation_step(self, batch_data, batch_idx): return self.shared_step(batch_data,"val")
    def test_step(self, batch_data, batch_idx):       return self.shared_step(batch_data,"test")

    def configure_optimizers(self):
        optimizer = torch.optim.AdamW(self.parameters(), lr=self.initial_lr, weight_decay=1e-5)

        # Расчет общего количества шагов для PolyLR
        # Опираемся на train_dataset
        # Добавляем проверку, что train_dataset доступен
        if not hasattr(self, 'train_dataset') or self.train_dataset is None or len(self.train_dataset) == 0:
             print("Warning: train_dataset not initialized or is empty. Cannot configure LR scheduler properly.")
             # Вернуть только оптимизатор или пропустить планировщик
             return {"optimizer": optimizer}

        num_train_samples = len(self.train_dataset)
        # Убедимся, что batch_size > 0 во избежание деления на ноль
        effective_batch_size = self.batch_size * self.trainer.accumulate_grad_batches if hasattr(self.trainer, 'accumulate_grad_batches') else self.batch_size
        if effective_batch_size <= 0: effective_batch_size = 1 # Запасной вариант

        num_batches_per_epoch = (num_train_samples + effective_batch_size - 1) // effective_batch_size # Целочисленное деление с округлением вверх

        total_training_steps = num_batches_per_epoch * self.num_total_epochs

        if total_training_steps <= 0:
             print("Warning: total_training_steps is zero or negative. Cannot configure LR scheduler properly.")
             return {"optimizer": optimizer}


        scheduler = torch.optim.lr_scheduler.LambdaLR(
            optimizer,
            lr_lambda=lambda step: poly_lr_scheduler(step, total_training_steps, 1.0) # 1.0 т.к. initial_lr уже в оптимайзере
        )
        return {"optimizer": optimizer, "lr_scheduler": {"scheduler": scheduler, "interval": "step"}}


# ================================================================
# 7. FUNCTION TO RUN A SINGLE EXPERIMENT
# ================================================================
def run_single_experiment(model_key, learning_rate, epochs_count, current_physical_batch_size, use_amp=False, grad_accum_steps=1):
    logger_to_use = (pl.loggers.WandbLogger(project="seg_drone_improved", name=f"{model_key}_bs{current_physical_batch_size}{'_amp' if use_amp else ''}{'_acc' + str(grad_accum_steps) if grad_accum_steps > 1 else ''}")
                     if USE_WANDB else True) # True for default Lightning logger

    model_instance = LitSegImproved(model_key_name=model_key, initial_lr=learning_rate,
                                   num_total_epochs=epochs_count, batch_size=current_physical_batch_size)

    trainer_config_params = {
        "accelerator": "gpu" if torch.cuda.is_available() else "cpu",
        "devices": 1,
        "max_epochs": epochs_count,
        "logger": logger_to_use,
        "log_every_n_steps": 10,
        "enable_checkpointing": False,
        "accumulate_grad_batches": grad_accum_steps # Передаем сюда накопление
    }

    if use_amp:
        trainer_config_params["precision"] = "16-mixed"

    # Убираем accumulate_grads из аргументов run_single_experiment
    # Он теперь напрямую передается в trainer_config_params

    trainer_instance = pl.Trainer(**trainer_config_params)

    print(f"\n--- Обучение улучшенной модели: {model_key} ---")
    # Используем current_physical_batch_size для отображения
    effective_batch_size = current_physical_batch_size * grad_accum_steps
    print(f"Параметры: LR={learning_rate}, Эпох={epochs_count}, Физ.батч={current_physical_batch_size}, Эфф.батч={effective_batch_size}, AMP={use_amp}, Разрешение={H_IMG}x{W_IMG}")

    # Запускаем обучение и тестирование
    trainer_instance.fit(model_instance)
    print(f"\n--- Тестирование улучшенной модели: {model_key} ---")
    trainer_instance.test(model_instance)

    del model_instance, trainer_instance # Освобождаем память
    gc.collect()
    torch.cuda.empty_cache()

# ================================================================
# 8. RUN EXPERIMENT FOR DeepLabV3Plus_mit_b2 (самые агрессивные настройки)
# ================================================================
EPOCHS_COUNT_DLV3 = 15 # Количество эпох
MODEL_KEY_DLV3 = "DeepLabV3Plus_mit_b2"
INITIAL_LR_DLV3 = 2e-4 # Начальный LR

print(f"\n{'='*15} ЗАПУСК УЛУЧШЕННОГО БЕЙЗЛАЙНА ДЛЯ {MODEL_KEY_DLV3} {'='*15}")

# Настройки для DeepLabV3+ mit_b2 с учетом ошибок и памяти
# Используем физический батч > 1 (например, 2), AMP, и Accumulation
# Эффективный батч = current_physical_batch_size * grad_accum_steps
run_single_experiment(
    model_key=MODEL_KEY_DLV3,
    learning_rate=INITIAL_LR_DLV3,
    epochs_count=EPOCHS_COUNT_DLV3,
    current_physical_batch_size=2, # ФИЗИЧЕСКИЙ БАТЧ > 1, например 2. Это обходит ошибку Batch=1, 1x1 spatial
    use_amp=True,                  # Mixed Precision (сильно экономит память)
    grad_accum_steps=4             # Накопление градиентов (эффективный батч = 2 * 4 = 8)
)

print(f"\n{'='*15} ОБУЧЕНИЕ МОДЕЛИ {MODEL_KEY_DLV3} ЗАВЕРШЕНО {'='*15}")

Локальная копия /content/dataset_local_binary_seg уже существует, используем ее.
Найдено 400 изображений.
Разбиение: train=320, val=40, test=40 семплов.

=============== ЗАПУСК УЛУЧШЕННОГО БЕЙЗЛАЙНА ДЛЯ DeepLabV3Plus_mit_b2 ===============


INFO:pytorch_lightning.utilities.rank_zero:Using 16bit Automatic Mixed Precision (AMP)
INFO:pytorch_lightning.utilities.rank_zero:GPU available: True (cuda), used: True
INFO:pytorch_lightning.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO:pytorch_lightning.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:pytorch_lightning.callbacks.model_summary:
  | Name  | Type          | Params | Mode 
------------------------------------------------
0 | model | DeepLabV3Plus | 25.3 M | train
------------------------------------------------
25.3 M    Trainable params
0         Non-trainable params
25.3 M    Total params
101.396   Total estimated model params size (MB)
383       Modules in train mode
0         Modules in eval mode



--- Обучение улучшенной модели: DeepLabV3Plus_mit_b2 ---
Параметры: LR=0.0002, Эпох=15, Физ.батч=2, Эфф.батч=8, AMP=True, Разрешение=736x736


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=15` reached.
INFO:pytorch_lightning.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]



--- Тестирование улучшенной модели: DeepLabV3Plus_mit_b2 ---


Testing: |          | 0/? [00:00<?, ?it/s]

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_acc          │    0.9283722639083862     │
│         test_iou          │     0.756996750831604     │
│         test_loss         │    0.1764976680278778     │
└───────────────────────────┴───────────────────────────┘


=============== ОБУЧЕНИЕ МОДЕЛИ DeepLabV3Plus_mit_b2 ЗАВЕРШЕНО ===============


Во втором коде для обучения тяжёлой модели DeepLabV3+ mit_b2 были внесены оптимизации: данные копируются на локальный диск для ускорения доступа, физический размер батча снижен до 2 для экономии памяти, добавлено накопление градиентов (accumulate_grad_batches=4) для увеличения эффективного размера батча, включена смешанная точность (AMP) для снижения потребления GPU-памяти, увеличено число потоков загрузки данных (num_workers=4), а также реализовано явное освобождение памяти после эксперимента — всё это позволило избежать ошибок Out of Memory и ValueError при обучении на ограниченных ресурсах Google Colab.

**Результаты**

**Модель 1: U-Net с энкодером ResNet50, улучшенный бейзлайн**  
- **Параметры:** 32.5 млн (увеличение на 33% от базовой U-Net ResNet34), Размер модели: ~130 МБ
- **Тестовые метрики:**  
  - mIoU = 0.732 (ниже, чем базовый U-Net ResNet34 — 0.756).  
  - Pixel Accuracy = 0.925 (близко к базовой модели).  
  - Loss = 0.202 (хуже, чем базовый 0.131).  

**Модель 2: DeepLabV3+ с энкодером mit_b2 (трансформер), улучшенный бейзлайн**  
- **Параметры:** 25.3 млн (увеличение на 565% от базового DeepLabV3+ mit_b0), Размер модели: ~101 МБ
- **Тестовые метрики:**  
  - mIoU = 0.757 (на 3% выше, чем базовый DeepLabV3+ mit_b0 — 0.727).  
  - Pixel Accuracy = 0.928 (на 0.002 выше базовой модели).  
  - Loss = 0.176 (значительно лучше базового 0.141).  


#### Выводы и сравнение


**Сравнение с базовым бейзлайном:**

| Метрика               | U-Net ResNet34 (база) | U-Net ResNet50 (улучш.) | DeepLabV3+ mit_b0 (база) | DeepLabV3+ mit_b2 (улучш.) |
|-----------------------|-----------------------|-------------------------|--------------------------|---------------------------|
| **mIoU (Test)**       | **0.756**             | 0.732 (↓3.2%)           | 0.727                   | **0.757 (↑4.1%)**         |
| **Pixel Accuracy**    | 0.933                | 0.925 (↓0.8%)           | 0.926                   | 0.928 (↑0.2%)             |
| **Loss (Test)**       | 0.131                | 0.202 (↑54%)            | 0.141                   | **0.176 (↓25%)**          |
| **Параметры**         | 24.4M                 | 32.5M (↑33%)            | 4.1M                    | 25.3M (↑565%)             |
| **Скорость обучения** | ~41 сек/эпоха         | Не указано              | ~55 сек/эпоха            | ~4 сек/шаг (эфф. батч=8)  |

**Ключевые выводы:**
1. **DeepLabV3+ mit_b2** превзошел базовую версию (mit_b0) по всем метрикам, демонстрируя преимущества более мощного трансформерного энкодера и улучшенных гиперпараметров (комбинированная функция потерь, PolyLR, накопление градиентов).  
2. **U-Net ResNet50** проиграл базовой U-Net ResNet34 по mIoU и Loss, что может быть связано с переобучением или избыточной сложностью модели.
3. **Скорость обучения** DeepLabV3+ mit_b2 (эффективный батч=8) улучшилась из-за оптимизации памяти (AMP, накопление градиентов), несмотря на большие параметры.  
4. **Трансформерные модели** (DeepLabV3+ mit_b2) показали более высокую стабильность: при увеличении параметров на 565% метрики выросли, тогда как сверточная U-Net ResNet50 ухудшилась.  
5. **Необходима оптимизация U-Net ResNet50:** Нужна перенастройка гиперпараметров, менее агрессивные аугментации, в целом более тонкая настройка для более сложного энкодера ResNet50.

####Самостоятельная реализация

#####Baseline

In [4]:
# ================================================================
# 1. ИМПОРТЫ
# ================================================================
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import numpy as np
import cv2
from pathlib import Path
from tqdm import tqdm
import albumentations as A
from albumentations.pytorch import ToTensorV2

# ================================================================
# 2. ПОДГОТОВКА ДАННЫХ
# ================================================================
# --- Параметры ---
DATA_ROOT = Path("/kaggle/input/semantic-segmentation-drone-dataset/binary_dataset/binary_dataset")
IMG_DIR = DATA_ROOT / "original_images"
MASK_DIR = DATA_ROOT / "images_semantic"
H, W = 512, 512

# --- Сбор файлов ---
imgs = sorted(set(list(IMG_DIR.glob("*.png")) + list(IMG_DIR.glob("*.jpg"))))
masks = [MASK_DIR / f"{p.stem}.png" for p in imgs]
pairs = [(i, m) for i, m in zip(imgs, masks) if m.exists()]

print(f"IMG_DIR: {IMG_DIR}, MASK_DIR: {MASK_DIR}")
print(f"Найдено изображений: {len(imgs)}")
print(f"Найдено пар изображение-маска: {len(pairs)}")
if len(pairs) == 0:
    raise RuntimeError("Не найдено ни одной пары изображение-маска! Проверьте имена файлов и структуру папок.")

imgs, masks = zip(*pairs)

# --- Разделение на train/val/test ---
idxs = np.random.permutation(len(imgs))
n = len(imgs)
n_train = int(0.8 * n)
n_val = int(0.1 * n)
split_files = {
    "train": (np.array(imgs)[idxs[:n_train]], np.array(masks)[idxs[:n_train]]),
    "val":   (np.array(imgs)[idxs[n_train:n_train+n_val]], np.array(masks)[idxs[n_train:n_train+n_val]]),
    "test":  (np.array(imgs)[idxs[n_train+n_val:]], np.array(masks)[idxs[n_train+n_val:]])
}

# ================================================================
# 3. АУГМЕНТАЦИИ
# ================================================================
def make_resize(h, w):
    return A.Resize(height=h, width=w)

def make_rrc(h, w, scale=(0.5, 1.0), p=1.0):
    return A.RandomResizedCrop(size=(h, w), scale=scale, p=p)

train_tf_improved = A.Compose([
    make_rrc(H, W, scale=(0.5, 1.0), p=1.0),
    A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.1, rotate_limit=15, p=0.7),
    A.HorizontalFlip(p=0.5),
    A.VerticalFlip(p=0.5),
    A.RandomBrightnessContrast(brightness_limit=0.3, contrast_limit=0.3, p=0.5),
    A.GaussianBlur(blur_limit=(3, 7), p=0.3),
    A.GaussNoise(var_limit=(10.0, 50.0), p=0.3),
    A.Normalize(),
    ToTensorV2(),
])

test_tf_improved = A.Compose([
    make_resize(H, W),
    A.Normalize(),
    ToTensorV2()
])

# ================================================================
# 4. КЛАСС ДАТАСЕТА
# ================================================================
class DroneBinDS(Dataset):
    def __init__(self, imgs_paths, masks_paths, transform_fn):
        self.imgs_paths = list(imgs_paths)
        self.masks_paths = list(masks_paths)
        self.transform_fn = transform_fn

    def __len__(self):
        return len(self.imgs_paths)

    def __getitem__(self, i):
        img = cv2.cvtColor(cv2.imread(str(self.imgs_paths[i])), cv2.COLOR_BGR2RGB)
        mask = cv2.imread(str(self.masks_paths[i]), 0)
        mask = (mask > 127).astype("float32")
        augmented = self.transform_fn(image=img, mask=mask)
        image_tensor = augmented["image"]
        mask_tensor = augmented["mask"]
        if mask_tensor.ndim == 2:
            mask_tensor = mask_tensor.unsqueeze(0)
        return image_tensor, mask_tensor.float()

# ================================================================
# 5. МЕТРИКИ
# ================================================================
def iou_score(pred, target, threshold=0.5, eps=1e-6):
    pred = (pred > threshold).float()
    target = (target > threshold).float()
    intersection = (pred * target).sum(dim=(1,2,3))
    union = (pred + target - pred * target).sum(dim=(1,2,3))
    iou = (intersection + eps) / (union + eps)
    return iou.mean().item()

def dice_score(pred, target, threshold=0.5, eps=1e-6):
    pred = (pred > threshold).float()
    target = (target > threshold).float()
    intersection = (pred * target).sum(dim=(1,2,3))
    dice = (2 * intersection + eps) / (pred.sum(dim=(1,2,3)) + target.sum(dim=(1,2,3)) + eps)
    return dice.mean().item()

def pixel_accuracy(pred, target, threshold=0.5):
    pred = (pred > threshold).float()
    target = (target > threshold).float()
    correct = (pred == target).float().sum()
    total = torch.numel(pred)
    return (correct / total).item()

# ================================================================
# 6. U-NET: САМОСТОЯТЕЛЬНАЯ РЕАЛИЗАЦИЯ
# ================================================================
class UNetBlock(nn.Module):
    def __init__(self, in_ch, out_ch):
        super().__init__()
        self.conv = nn.Sequential(
            nn.Conv2d(in_ch, out_ch, 3, padding=1),
            nn.BatchNorm2d(out_ch),
            nn.ReLU(inplace=True),
            nn.Conv2d(out_ch, out_ch, 3, padding=1),
            nn.BatchNorm2d(out_ch),
            nn.ReLU(inplace=True)
        )

    def forward(self, x):
        return self.conv(x)

class UNet(nn.Module):
    def __init__(self, in_ch=3, out_ch=1, features=[64, 128, 256, 512]):
        super().__init__()
        self.downs = nn.ModuleList()
        self.ups = nn.ModuleList()
        for feat in features:
            self.downs.append(UNetBlock(in_ch, feat))
            in_ch = feat
        self.bottleneck = UNetBlock(features[-1], features[-1]*2)
        for feat in reversed(features):
            self.ups.append(nn.ConvTranspose2d(feat*2, feat, 2, stride=2))
            self.ups.append(UNetBlock(feat*2, feat))
        self.final = nn.Conv2d(features[0], out_ch, 1)

    def forward(self, x):
        skips = []
        for down in self.downs:
            x = down(x)
            skips.append(x)
            x = F.max_pool2d(x, 2)
        x = self.bottleneck(x)
        skips = skips[::-1]
        for i in range(0, len(self.ups), 2):
            x = self.ups[i](x)
            skip = skips[i//2]
            if x.shape != skip.shape:
                x = F.interpolate(x, size=skip.shape[2:])
            x = torch.cat([skip, x], dim=1)
            x = self.ups[i+1](x)
        return self.final(x)

# ================================================================
# 7. ОБУЧАЮЩИЙ И ВАЛИДАЦИОННЫЙ ЦИКЛ
# ================================================================
def train_one_epoch(model, loader, optimizer, loss_fn, device):
    model.train()
    total_loss = 0
    for imgs, masks in tqdm(loader, desc="Train", leave=False):
        imgs, masks = imgs.to(device), masks.to(device)
        optimizer.zero_grad()
        # === Mixed Precision (AMP) для экономии памяти ===
        with torch.cuda.amp.autocast():
            logits = model(imgs)
            loss = loss_fn(logits, masks)
        scaler.scale(loss).backward()
        scaler.step(optimizer)
        scaler.update()
        total_loss += loss.item()
        # === Очистка кэша после каждого батча (дополнительно) ===
        torch.cuda.empty_cache()
    return total_loss / len(loader)

def evaluate(model, loader, device):
    model.eval()
    iou, dice, acc, loss_sum = 0, 0, 0, 0
    loss_fn = nn.BCEWithLogitsLoss()
    with torch.no_grad():
        for imgs, masks in tqdm(loader, desc="Val", leave=False):
            imgs, masks = imgs.to(device), masks.to(device)
            logits = model(imgs)
            probs = torch.sigmoid(logits)
            loss = loss_fn(logits, masks)
            iou += iou_score(probs, masks)
            dice += dice_score(probs, masks)
            acc += pixel_accuracy(probs, masks)
            loss_sum += loss.item()
    n = len(loader)
    return {
        "iou": iou/n,
        "dice": dice/n,
        "acc": acc/n,
        "loss": loss_sum/n
    }

# ================================================================
# 8. ОБУЧЕНИЕ И ТЕСТИРОВАНИЕ
# ================================================================
BATCH_SIZE = 2
NUM_WORKERS = 2
EPOCHS = 20

train_ds = DroneBinDS(*split_files["train"], train_tf_improved)
val_ds   = DroneBinDS(*split_files["val"],   test_tf_improved)
test_ds  = DroneBinDS(*split_files["test"],  test_tf_improved)

train_loader = DataLoader(train_ds, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
val_loader   = DataLoader(val_ds,   batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)
test_loader  = DataLoader(test_ds,  batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = UNet(in_ch=3, out_ch=1).to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
loss_fn = nn.BCEWithLogitsLoss()

# === Mixed Precision (AMP) scaler ===
scaler = torch.cuda.amp.GradScaler()

best_val_iou = 0
best_model_state = None

for epoch in range(EPOCHS):
    print(f"\nEpoch {epoch+1}/{EPOCHS}")
    torch.cuda.empty_cache()  # Очистка памяти перед эпохой
    train_loss = train_one_epoch(model, train_loader, optimizer, loss_fn, device)
    val_metrics = evaluate(model, val_loader, device)
    print(f"Train Loss: {train_loss:.4f} | Val Loss: {val_metrics['loss']:.4f} | Val IoU: {val_metrics['iou']:.4f} | Val Dice: {val_metrics['dice']:.4f} | Val Acc: {val_metrics['acc']:.4f}")
    if val_metrics['iou'] > best_val_iou:
        best_val_iou = val_metrics['iou']
        best_model_state = model.state_dict()

print("\n=== Оценка на тестовой выборке ===")
model.load_state_dict(best_model_state)
test_metrics = evaluate(model, test_loader, device)
print(f"Test Loss: {test_metrics['loss']:.4f} | Test IoU: {test_metrics['iou']:.4f} | Test Dice: {test_metrics['dice']:.4f} | Test Acc: {test_metrics['acc']:.4f}")

IMG_DIR: /kaggle/input/semantic-segmentation-drone-dataset/binary_dataset/binary_dataset/original_images, MASK_DIR: /kaggle/input/semantic-segmentation-drone-dataset/binary_dataset/binary_dataset/images_semantic
Найдено изображений: 400
Найдено пар изображение-маска: 400


/usr/local/lib/python3.11/dist-packages/albumentations/core/validation.py:111: UserWarning: ShiftScaleRotate is a special case of Affine transform. Please use Affine transform instead.
  original_init(self, **validated_kwargs)
<ipython-input-4-6999ccb1e9a5>:64: UserWarning: Argument(s) 'var_limit' are not valid for transform GaussNoise
  A.GaussNoise(var_limit=(10.0, 50.0), p=0.3),
<ipython-input-4-6999ccb1e9a5>:236: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()



Epoch 1/20


Train:   0%|          | 0/160 [00:00<?, ?it/s]<ipython-input-4-6999ccb1e9a5>:182: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast():


Train Loss: 0.5124 | Val Loss: 0.4630 | Val IoU: 0.0002 | Val Dice: 0.0005 | Val Acc: 0.7784

Epoch 2/20


Train Loss: 0.4722 | Val Loss: 0.4633 | Val IoU: 0.0393 | Val Dice: 0.0721 | Val Acc: 0.7855

Epoch 3/20


Train Loss: 0.4651 | Val Loss: 0.4565 | Val IoU: 0.1140 | Val Dice: 0.1944 | Val Acc: 0.7620

Epoch 4/20


Train Loss: 0.4417 | Val Loss: 0.4316 | Val IoU: 0.0735 | Val Dice: 0.1309 | Val Acc: 0.7769

Epoch 5/20


Train Loss: 0.4483 | Val Loss: 0.4617 | Val IoU: 0.2914 | Val Dice: 0.4268 | Val Acc: 0.7871

Epoch 6/20


Train Loss: 0.4322 | Val Loss: 0.4069 | Val IoU: 0.1486 | Val Dice: 0.2449 | Val Acc: 0.7979

Epoch 7/20


Train Loss: 0.4264 | Val Loss: 0.4464 | Val IoU: 0.2518 | Val Dice: 0.3835 | Val Acc: 0.7877

Epoch 8/20


Train Loss: 0.4308 | Val Loss: 0.4359 | Val IoU: 0.0997 | Val Dice: 0.1666 | Val Acc: 0.7840

Epoch 9/20


Train Loss: 0.4141 | Val Loss: 0.4390 | Val IoU: 0.3891 | Val Dice: 0.5358 | Val Acc: 0.7966

Epoch 10/20


Train Loss: 0.4212 | Val Loss: 0.4457 | Val IoU: 0.3906 | Val Dice: 0.5392 | Val Acc: 0.7740

Epoch 11/20


Train Loss: 0.4105 | Val Loss: 0.3953 | Val IoU: 0.3724 | Val Dice: 0.5198 | Val Acc: 0.8075

Epoch 12/20


Train Loss: 0.4015 | Val Loss: 0.5496 | Val IoU: 0.3956 | Val Dice: 0.5454 | Val Acc: 0.7482

Epoch 13/20


Train Loss: 0.4085 | Val Loss: 0.4103 | Val IoU: 0.3404 | Val Dice: 0.4869 | Val Acc: 0.8208

Epoch 14/20


Train Loss: 0.4018 | Val Loss: 0.4051 | Val IoU: 0.1994 | Val Dice: 0.3161 | Val Acc: 0.8028

Epoch 15/20


Train Loss: 0.3914 | Val Loss: 0.4591 | Val IoU: 0.2330 | Val Dice: 0.3591 | Val Acc: 0.7948

Epoch 16/20


Train Loss: 0.3956 | Val Loss: 0.3978 | Val IoU: 0.3706 | Val Dice: 0.5192 | Val Acc: 0.8099

Epoch 17/20


Train Loss: 0.4034 | Val Loss: 0.4577 | Val IoU: 0.2140 | Val Dice: 0.3353 | Val Acc: 0.7986

Epoch 18/20


Train Loss: 0.3987 | Val Loss: 0.3826 | Val IoU: 0.3314 | Val Dice: 0.4738 | Val Acc: 0.8259

Epoch 19/20


Train Loss: 0.3987 | Val Loss: 0.4443 | Val IoU: 0.2163 | Val Dice: 0.3403 | Val Acc: 0.7883

Epoch 20/20


Train Loss: 0.3905 | Val Loss: 0.4081 | Val IoU: 0.3831 | Val Dice: 0.5320 | Val Acc: 0.8124

=== Оценка на тестовой выборке ===


Test Loss: 0.3690 | Test IoU: 0.3656 | Test Dice: 0.5143 | Test Acc: 0.8303


В процессе снова была получена ошибка OutOfMemoryError: CUDA out of memory

Что использовано для экономии памяти:
* Уменьшен размер изображений (H, W = 512, 512)
* Уменьшен BATCH_SIZE = 2
* Уменьшен NUM_WORKERS = 2 и pin_memory=False
* Добавлен AMP (Mixed Precision) через torch.cuda.amp.autocast() и GradScaler
* После каждого батча и эпохи вызывается torch.cuda.empty_cache()

Качество сегментации собственной реализации U-Net существенно ниже, чем у библиотечных моделей SMP.
Разница по IoU составляет более чем в два раза (0.366 против 0.756 у U-Net ResNet34).
Точность по пикселям и Dice также заметно ниже.
Loss на тесте выше, что говорит о менее уверенных предсказаниях.
Это ожидаемо, так как библиотечные реализации используют более мощные энкодеры (ResNet34/50, MiT), предобученные веса и более продвинутые техники оптимизации.

Причины разницы:
В собственной реализации используется "чистый" U-Net без предобученного энкодера и без глубоких архитектурных улучшений.
В библиотечных моделях применяются современные энкодеры, предобученные на ImageNet, что даёт значительный прирост качества.
Также в SMP-реализациях используются более сложные функции потерь и scheduler'ы.

#####Улучшение Baseline

Главные улучшения:

* Более сильные аугментации (пространственные + цветовые)
* Комбинированная функция потерь (Dice + BCE)
* AdamW + PolyLR (реализация вручную)
* AMP (Mixed Precision)
* Всё оптимизировано под экономию памяти (batch=2, H,W=512, torch.cuda.empty_cache())

In [5]:
# ================================================================
# 1. ИМПОРТЫ
# ================================================================
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import numpy as np
import cv2
from pathlib import Path
from tqdm import tqdm
import albumentations as A
from albumentations.pytorch import ToTensorV2

# ================================================================
# 2. ПОДГОТОВКА ДАННЫХ
# ================================================================
DATA_ROOT = Path("/kaggle/input/semantic-segmentation-drone-dataset/binary_dataset/binary_dataset")
IMG_DIR = DATA_ROOT / "original_images"
MASK_DIR = DATA_ROOT / "images_semantic"
H, W = 512, 512

imgs = sorted(set(list(IMG_DIR.glob("*.png")) + list(IMG_DIR.glob("*.jpg"))))
masks = [MASK_DIR / f"{p.stem}.png" for p in imgs]
pairs = [(i, m) for i, m in zip(imgs, masks) if m.exists()]

print(f"IMG_DIR: {IMG_DIR}, MASK_DIR: {MASK_DIR}")
print(f"Найдено изображений: {len(imgs)}")
print(f"Найдено пар изображение-маска: {len(pairs)}")
if len(pairs) == 0:
    raise RuntimeError("Не найдено ни одной пары изображение-маска! Проверьте имена файлов и структуру папок.")

imgs, masks = zip(*pairs)

idxs = np.random.permutation(len(imgs))
n = len(imgs)
n_train = int(0.8 * n)
n_val = int(0.1 * n)
split_files = {
    "train": (np.array(imgs)[idxs[:n_train]], np.array(masks)[idxs[:n_train]]),
    "val":   (np.array(imgs)[idxs[n_train:n_train+n_val]], np.array(masks)[idxs[n_train:n_train+n_val]]),
    "test":  (np.array(imgs)[idxs[n_train+n_val:]], np.array(masks)[idxs[n_train+n_val:]])
}

# ================================================================
# 3. АУГМЕНТАЦИИ (ещё сильнее!)
# ================================================================
def make_resize(h, w):
    return A.Resize(height=h, width=w)

def make_rrc(h, w, scale=(0.5, 1.0), p=1.0):
    return A.RandomResizedCrop(size=(h, w), scale=scale, p=p)

train_tf_improved = A.Compose([
    make_rrc(H, W, scale=(0.5, 1.0), p=1.0),
    A.ShiftScaleRotate(shift_limit=0.1, scale_limit=0.15, rotate_limit=20, p=0.8),
    A.HorizontalFlip(p=0.5),
    A.VerticalFlip(p=0.5),
    A.RandomBrightnessContrast(brightness_limit=0.4, contrast_limit=0.4, p=0.7),
    A.CLAHE(p=0.3),
    A.RandomGamma(p=0.3),
    A.GaussianBlur(blur_limit=(3, 7), p=0.4),
    A.GaussNoise(var_limit=(10.0, 50.0), p=0.4),
    A.GridDistortion(p=0.2),
    A.CoarseDropout(max_holes=8, max_height=32, max_width=32, fill_value=0, p=0.3),
    A.Normalize(),
    ToTensorV2(),
])

test_tf_improved = A.Compose([
    make_resize(H, W),
    A.Normalize(),
    ToTensorV2()
])

# ================================================================
# 4. КЛАСС ДАТАСЕТА
# ================================================================
class DroneBinDS(Dataset):
    def __init__(self, imgs_paths, masks_paths, transform_fn):
        self.imgs_paths = list(imgs_paths)
        self.masks_paths = list(masks_paths)
        self.transform_fn = transform_fn

    def __len__(self):
        return len(self.imgs_paths)

    def __getitem__(self, i):
        img = cv2.cvtColor(cv2.imread(str(self.imgs_paths[i])), cv2.COLOR_BGR2RGB)
        mask = cv2.imread(str(self.masks_paths[i]), 0)
        mask = (mask > 127).astype("float32")
        augmented = self.transform_fn(image=img, mask=mask)
        image_tensor = augmented["image"]
        mask_tensor = augmented["mask"]
        if mask_tensor.ndim == 2:
            mask_tensor = mask_tensor.unsqueeze(0)
        return image_tensor, mask_tensor.float()

# ================================================================
# 5. МЕТРИКИ
# ================================================================
def iou_score(pred, target, threshold=0.5, eps=1e-6):
    pred = (pred > threshold).float()
    target = (target > threshold).float()
    intersection = (pred * target).sum(dim=(1,2,3))
    union = (pred + target - pred * target).sum(dim=(1,2,3))
    iou = (intersection + eps) / (union + eps)
    return iou.mean().item()

def dice_score(pred, target, threshold=0.5, eps=1e-6):
    pred = (pred > threshold).float()
    target = (target > threshold).float()
    intersection = (pred * target).sum(dim=(1,2,3))
    dice = (2 * intersection + eps) / (pred.sum(dim=(1,2,3)) + target.sum(dim=(1,2,3)) + eps)
    return dice.mean().item()

def pixel_accuracy(pred, target, threshold=0.5):
    pred = (pred > threshold).float()
    target = (target > threshold).float()
    correct = (pred == target).float().sum()
    total = torch.numel(pred)
    return (correct / total).item()

# ================================================================
# 6. U-NET: САМОСТОЯТЕЛЬНАЯ РЕАЛИЗАЦИЯ
# ================================================================
class UNetBlock(nn.Module):
    def __init__(self, in_ch, out_ch):
        super().__init__()
        self.conv = nn.Sequential(
            nn.Conv2d(in_ch, out_ch, 3, padding=1),
            nn.BatchNorm2d(out_ch),
            nn.ReLU(inplace=True),
            nn.Conv2d(out_ch, out_ch, 3, padding=1),
            nn.BatchNorm2d(out_ch),
            nn.ReLU(inplace=True)
        )

    def forward(self, x):
        return self.conv(x)

class UNet(nn.Module):
    def __init__(self, in_ch=3, out_ch=1, features=[64, 128, 256, 512]):
        super().__init__()
        self.downs = nn.ModuleList()
        self.ups = nn.ModuleList()
        for feat in features:
            self.downs.append(UNetBlock(in_ch, feat))
            in_ch = feat
        self.bottleneck = UNetBlock(features[-1], features[-1]*2)
        for feat in reversed(features):
            self.ups.append(nn.ConvTranspose2d(feat*2, feat, 2, stride=2))
            self.ups.append(UNetBlock(feat*2, feat))
        self.final = nn.Conv2d(features[0], out_ch, 1)

    def forward(self, x):
        skips = []
        for down in self.downs:
            x = down(x)
            skips.append(x)
            x = F.max_pool2d(x, 2)
        x = self.bottleneck(x)
        skips = skips[::-1]
        for i in range(0, len(self.ups), 2):
            x = self.ups[i](x)
            skip = skips[i//2]
            if x.shape != skip.shape:
                x = F.interpolate(x, size=skip.shape[2:])
            x = torch.cat([skip, x], dim=1)
            x = self.ups[i+1](x)
        return self.final(x)

# ================================================================
# 7. КОМБИНИРОВАННЫЙ ЛОСС (Dice + BCE)
# ================================================================
class DiceLoss(nn.Module):
    def __init__(self, eps=1e-6):
        super().__init__()
        self.eps = eps
    def forward(self, logits, targets):
        probs = torch.sigmoid(logits)
        targets = (targets > 0.5).float()
        intersection = (probs * targets).sum(dim=(1,2,3))
        union = probs.sum(dim=(1,2,3)) + targets.sum(dim=(1,2,3))
        dice = (2 * intersection + self.eps) / (union + self.eps)
        return 1 - dice.mean()

dice_loss = DiceLoss()
bce_loss = nn.BCEWithLogitsLoss()
def combo_loss_fn(logits, targets):
    return 0.7 * dice_loss(logits, targets) + 0.3 * bce_loss(logits, targets)

# ================================================================
# 8. PolyLR SCHEDULER
# ================================================================
def poly_lr_scheduler(optimizer, init_lr, curr_iter, max_iter, power=0.9):
    lr = init_lr * (1 - curr_iter / max_iter) ** power
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr

# ================================================================
# 9. ОБУЧАЮЩИЙ И ВАЛИДАЦИОННЫЙ ЦИКЛ (AMP + PolyLR)
# ================================================================
def train_one_epoch(model, loader, optimizer, loss_fn, device, scaler, epoch, max_iter, init_lr):
    model.train()
    total_loss = 0
    for batch_idx, (imgs, masks) in enumerate(tqdm(loader, desc="Train", leave=False)):
        imgs, masks = imgs.to(device), masks.to(device)
        optimizer.zero_grad()
        with torch.cuda.amp.autocast():
            logits = model(imgs)
            loss = loss_fn(logits, masks)
        scaler.scale(loss).backward()
        scaler.step(optimizer)
        scaler.update()
        # PolyLR step
        poly_lr_scheduler(optimizer, init_lr, epoch * len(loader) + batch_idx, max_iter)
        total_loss += loss.item()
        torch.cuda.empty_cache()
    return total_loss / len(loader)

def evaluate(model, loader, device):
    model.eval()
    iou, dice, acc, loss_sum = 0, 0, 0, 0
    loss_fn = combo_loss_fn
    with torch.no_grad():
        for imgs, masks in tqdm(loader, desc="Val", leave=False):
            imgs, masks = imgs.to(device), masks.to(device)
            logits = model(imgs)
            probs = torch.sigmoid(logits)
            loss = loss_fn(logits, masks)
            iou += iou_score(probs, masks)
            dice += dice_score(probs, masks)
            acc += pixel_accuracy(probs, masks)
            loss_sum += loss.item()
    n = len(loader)
    return {
        "iou": iou/n,
        "dice": dice/n,
        "acc": acc/n,
        "loss": loss_sum/n
    }

# ================================================================
# 10. ОБУЧЕНИЕ И ТЕСТИРОВАНИЕ
# ================================================================
BATCH_SIZE = 2
NUM_WORKERS = 2
EPOCHS = 20
INIT_LR = 1e-3

train_ds = DroneBinDS(*split_files["train"], train_tf_improved)
val_ds   = DroneBinDS(*split_files["val"],   test_tf_improved)
test_ds  = DroneBinDS(*split_files["test"],  test_tf_improved)

train_loader = DataLoader(train_ds, batch_size=BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS, pin_memory=True)
val_loader   = DataLoader(val_ds,   batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)
test_loader  = DataLoader(test_ds,  batch_size=BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS, pin_memory=True)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = UNet(in_ch=3, out_ch=1).to(device)
optimizer = torch.optim.AdamW(model.parameters(), lr=INIT_LR, weight_decay=1e-5)
scaler = torch.cuda.amp.GradScaler()

max_iter = EPOCHS * len(train_loader)
best_val_iou = 0
best_model_state = None

for epoch in range(EPOCHS):
    print(f"\nEpoch {epoch+1}/{EPOCHS}")
    torch.cuda.empty_cache()
    train_loss = train_one_epoch(model, train_loader, optimizer, combo_loss_fn, device, scaler, epoch, max_iter, INIT_LR)
    val_metrics = evaluate(model, val_loader, device)
    print(f"Train Loss: {train_loss:.4f} | Val Loss: {val_metrics['loss']:.4f} | Val IoU: {val_metrics['iou']:.4f} | Val Dice: {val_metrics['dice']:.4f} | Val Acc: {val_metrics['acc']:.4f}")
    if val_metrics['iou'] > best_val_iou:
        best_val_iou = val_metrics['iou']
        best_model_state = model.state_dict()

print("\n=== Оценка на тестовой выборке ===")
model.load_state_dict(best_model_state)
test_metrics = evaluate(model, test_loader, device)
print(f"Test Loss: {test_metrics['loss']:.4f} | Test IoU: {test_metrics['iou']:.4f} | Test Dice: {test_metrics['dice']:.4f} | Test Acc: {test_metrics['acc']:.4f}")

IMG_DIR: /kaggle/input/semantic-segmentation-drone-dataset/binary_dataset/binary_dataset/original_images, MASK_DIR: /kaggle/input/semantic-segmentation-drone-dataset/binary_dataset/binary_dataset/images_semantic
Найдено изображений: 400
Найдено пар изображение-маска: 400


/usr/local/lib/python3.11/dist-packages/albumentations/core/validation.py:111: UserWarning: ShiftScaleRotate is a special case of Affine transform. Please use Affine transform instead.
  original_init(self, **validated_kwargs)
<ipython-input-5-8e1048d6b9f9>:63: UserWarning: Argument(s) 'var_limit' are not valid for transform GaussNoise
  A.GaussNoise(var_limit=(10.0, 50.0), p=0.4),
<ipython-input-5-8e1048d6b9f9>:65: UserWarning: Argument(s) 'max_holes, max_height, max_width, fill_value' are not valid for transform CoarseDropout
  A.CoarseDropout(max_holes=8, max_height=32, max_width=32, fill_value=0, p=0.3),
<ipython-input-5-8e1048d6b9f9>:263: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()



Epoch 1/20


Train:   0%|          | 0/160 [00:00<?, ?it/s]<ipython-input-5-8e1048d6b9f9>:210: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast():


Train Loss: 0.6722 | Val Loss: 0.6163 | Val IoU: 0.2034 | Val Dice: 0.3231 | Val Acc: 0.7387

Epoch 2/20


Train Loss: 0.6499 | Val Loss: 0.6034 | Val IoU: 0.1682 | Val Dice: 0.2707 | Val Acc: 0.7502

Epoch 3/20


Train Loss: 0.6502 | Val Loss: 0.6053 | Val IoU: 0.2124 | Val Dice: 0.3313 | Val Acc: 0.7716

Epoch 4/20


Train Loss: 0.6256 | Val Loss: 0.6651 | Val IoU: 0.1983 | Val Dice: 0.3130 | Val Acc: 0.7261

Epoch 5/20


Train Loss: 0.6263 | Val Loss: 0.5714 | Val IoU: 0.2884 | Val Dice: 0.4283 | Val Acc: 0.7603

Epoch 6/20


Train Loss: 0.6091 | Val Loss: 0.6749 | Val IoU: 0.3486 | Val Dice: 0.4891 | Val Acc: 0.6786

Epoch 7/20


Train Loss: 0.6087 | Val Loss: 0.6001 | Val IoU: 0.2830 | Val Dice: 0.4154 | Val Acc: 0.7530

Epoch 8/20


Train Loss: 0.5960 | Val Loss: 0.5761 | Val IoU: 0.2751 | Val Dice: 0.4116 | Val Acc: 0.7885

Epoch 9/20


Train Loss: 0.5842 | Val Loss: 0.5937 | Val IoU: 0.3103 | Val Dice: 0.4479 | Val Acc: 0.7452

Epoch 10/20


Train Loss: 0.5884 | Val Loss: 0.5910 | Val IoU: 0.3364 | Val Dice: 0.4801 | Val Acc: 0.7584

Epoch 11/20


Train Loss: 0.5706 | Val Loss: 0.5647 | Val IoU: 0.3087 | Val Dice: 0.4470 | Val Acc: 0.7882

Epoch 12/20


Train Loss: 0.5623 | Val Loss: 0.5962 | Val IoU: 0.2699 | Val Dice: 0.4035 | Val Acc: 0.7756

Epoch 13/20


Train Loss: 0.5636 | Val Loss: 0.5887 | Val IoU: 0.2791 | Val Dice: 0.4162 | Val Acc: 0.7691

Epoch 14/20


Train Loss: 0.5594 | Val Loss: 0.5809 | Val IoU: 0.3185 | Val Dice: 0.4609 | Val Acc: 0.7646

Epoch 15/20


Train Loss: 0.5451 | Val Loss: 0.5849 | Val IoU: 0.2784 | Val Dice: 0.4132 | Val Acc: 0.7781

Epoch 16/20


Train Loss: 0.5580 | Val Loss: 0.5400 | Val IoU: 0.3545 | Val Dice: 0.5084 | Val Acc: 0.7708

Epoch 17/20


Train Loss: 0.5450 | Val Loss: 0.5537 | Val IoU: 0.3619 | Val Dice: 0.5062 | Val Acc: 0.7704

Epoch 18/20


Train Loss: 0.5391 | Val Loss: 0.5610 | Val IoU: 0.3568 | Val Dice: 0.4988 | Val Acc: 0.7700

Epoch 19/20


Train Loss: 0.5447 | Val Loss: 0.5385 | Val IoU: 0.3436 | Val Dice: 0.4921 | Val Acc: 0.7818

Epoch 20/20


Train Loss: 0.5295 | Val Loss: 0.5495 | Val IoU: 0.3336 | Val Dice: 0.4775 | Val Acc: 0.7786

=== Оценка на тестовой выборке ===


Test Loss: 0.4921 | Test IoU: 0.4142 | Test Dice: 0.5728 | Test Acc: 0.8016


Результаты улучшенного собственного U-Net:

* Test IoU (mIoU): 0.414
* Test Dice: 0.573
* Test Pixel Accuracy: 0.802
* Test Loss: 0.492

Сравнение с собственной базовой реализацией:

Улучшенный бейзлайн показывает заметный прирост по всем метрикам по сравнению с базовой собственной реализацией (где mIoU ≈ 0.366, Dice ≈ 0.514, Acc ≈ 0.83, Loss ≈ 0.369).
mIoU вырос примерно на 13%, Dice — на 6%, что говорит о положительном влиянии аугментаций, комбинированного лосса и других улучшений.

#### Выводы


| Модель/Реализация         | mIoU (Test) | Dice (Test) | Pixel Accuracy (Test) | Loss (Test) | Примечания                |
|---------------------------|-------------|-------------|-----------------------|-------------|---------------------------|
| **U-Net (self, базовый)** |   0.366     |   0.514     |        0.830          |   0.369     | Самостоятельная реализация, базовая |
| **U-Net (self, улучш.)**  |   0.414     |   0.573     |        0.802          |   0.492     | Самостоятельная реализация, улучшенная |
| U-Net ResNet34 (SMP)      |   0.756     |   0.86     |        0.933          |   0.131     | Библиотечная, базовая     |
| U-Net ResNet50 (SMP)      |   0.732     |   0.85     |        0.925          |   0.202     | Библиотечная, улучшенная  |
| DeepLabV3+ mit_b0 (SMP)   |   0.727     |   0.85     |        0.926          |   0.141     | Библиотечная, базовая     |
| DeepLabV3+ mit_b2 (SMP)   |   0.757     |   0.87     |        0.928          |   0.176     | Библиотечная, улучшенная  |